---
title: "Part 1 - Variable SNV Position and Visualization"
author: "Jörg Wennmann"
date: "2024-12-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Required libraries

```{r}
library(bacsnp)
library(vcfR)
library(patchwork)

#'%ni%' <- Negate('%in%')
```

## Input the VCF file

Input VCF and transform to readable dataframe using bacsnp transformation.

```{r}
data <- vcfR::read.vcfR("vcf/bmnpv_my_ja_vs_refIndia.vcf")
df <- bacsnp.transformation(data)

#Change NCBI SRR to BmNPV isolate description
df$ISO[df$ISO == "iSRR14313721"] <- "iJa"
df$ISO[df$ISO == "iSRR14313723"] <- "iMy"
```

## Determining SNP specificities

Tool bacsnp is used to determine SNP specificities.

```{r}
iso.spec <- c("iMy", "iJa")
df.spec <- bacsnp.specificity(df, isolates = iso.spec, which.rel = "REL.ALT1")

snp.data <- df.spec$data

snp.data <- snp.data[!is.na(snp.data$REL.ALT1),]

snp.data$id <- "in"

snp.data$ISO <- as.factor(snp.data$ISO)
snp.data$ISO <- factor(snp.data$ISO, levels = c("iMy", "iJa"))

snp.data$id <- as.factor(snp.data$id)
snp.data$id <- factor(snp.data$id, levels = c("in"))

snp.data$SPEC <- as.factor(snp.data$SPEC)
snp.data$SPEC <- factor(snp.data$SPEC, levels = c("Ja_My", "Ja", "My"))

#snp.data <- snp.data[order(factor(snp.data$SPEC)),]

myData <- subset(snp.data, ISO == "iMy")
jaData <- subset(snp.data, ISO == "iJa")
```


## Variable SNV position visualization 

```{r}
snv.plot <- function(x, col = "blue", genome.length = NULL, dotSize = 2.5, ALPHA = 0.6) {
  # Entferne Zeilen mit NA in REL.ALT1
  x <- x[!is.na(x$REL.ALT1), ]
  
  # Prüfe, ob `col` eine Spalte ist oder eine feste Farbe
  if (!col %in% names(x)) {
    x$color <- col  # Erstelle eine neue Spalte 'color' mit der festen Farbe
    color_layer <- geom_point(
      aes(x = as.numeric(POS), y = REL.ALT1),
      color = col, alpha = ALPHA, size = dotSize
    )  # Verwende die feste Farbe
  } else {
    color_layer <- geom_point(
      aes(x = as.numeric(POS), y = REL.ALT1, color = .data[[col]]),
      alpha = ALPHA, size = dotSize
    )  # Verwende die existierende Spalte für das Farb-Mapping
  }
  
  # Bestimme die Länge des Genoms und erstelle x-Achsen-Breaks
  if (is.numeric(genome.length)) {
    b <- seq(0, genome.length - 5000, by = 5000)
    b[1] <- 1
    b <- append(b, genome.length)
  } else {
    genome.length <- max(x$POS)
    b <- seq(0, genome.length, by = 5000)
    b[1] <- 1
    b <- append(b, genome.length)
  }
  
  # Erstelle den ggplot
  p <- ggplot(x) +
    theme_bw() +
    scale_x_continuous(
      limits = c(0, genome.length),
      expand = c(0.01, 0),
      breaks = b,
      labels = floor(b / 1000)
    ) +
    scale_y_continuous(
      limits = c(0, 1),
      breaks = seq(0, 1, 0.25)
    ) +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 0.5, vjust = 0.5),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(colour = "black"),
      panel.background = element_rect(fill = "white"),
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 1),
      legend.position = "bottom"
    ) +
    color_layer +  # Füge den richtigen Layer hinzu
    xlab("Genome position (kbp)") +
    ylab("Relative SNV Frequency")
  
  return(p)
}

a <- snv.plot(myData, col = "purple")
a <- a + labs(title = "BmNPV-My")
b <- snv.plot(myData, col = "SPEC")
b <- b + scale_color_discrete(name = "SNV specificities")
c <- snv.plot(jaData, col = "purple")
c <- c + labs(title = "BmNPV-Ja")
d <- snv.plot(jaData, col = "SPEC")
d <- d + scale_color_discrete(name = "SNV specificities")

## combined plot with patchwork
combined_plot <- (a | c) / (b | d) +
  plot_annotation(title = '') +
  plot_annotation(tag_levels = 'A')

print(combined_plot)

f <- 0.8

w <- 15
h <- 7

ggsave("figures/bacsnp_plot.png", plot = combined_plot, width = w * f, height = h * f) 
```

# Creating a BED file from SNV positions

```{r}
bed_data <- data.frame(
  chrom = snp.data$CHROM,         # Chromosome
  start = snp.data$POS - 1,       # 0-based start position
  end = snp.data$POS,             # 1-based end position
  name = snp.data$SPEC,           # Optional: Name of the region
  score = ".",                    # Optional: Score, here "." (no data)
  strand = "."                    # Optional: Strand, here "." (unknown)
)

write.table(
  bed_data,
  file = "bed/snv_data.bed",
  sep = "\t",
  quote = FALSE,
  row.names = FALSE,
  col.names = FALSE
)
```
