---
title: "Part 2 - Nanopore reads"
author: "Jörg Wennmann"
date: "2024-12-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Required libraries

```{r}
library(data.table)
library(ggplot2)
library(patchwork)
```

# Function for VCF to Matrix transformation

```{r}
process_vcf_to_snp_matrix <- function(vcf_file_path) {

  split_string_to_chars <- function(input_string) {
    strsplit(input_string, "")[[1]]
  }

  process_plus_minus_in_vector <- function(char_vector) {
    i <- 1
    processed_vector <- c()
    while (i <= length(char_vector)) {
      if (char_vector[i] %in% c("+", "-")) {
        if ((i + 1) <= length(char_vector) && grepl("^[0-9]+$", char_vector[i + 1])) {
          num <- as.numeric(char_vector[i + 1])
          i <- i + 1 + num
        } else {
          processed_vector <- c(processed_vector, char_vector[i])
        }
      } else {
        processed_vector <- c(processed_vector, char_vector[i])
      }
      i <- i + 1
    }
    processed_vector
  }

  replace_asterisk_and_hash <- function(char_vector) {
    gsub("\\*|#", "-", char_vector)
  }

  remove_caret_and_next <- function(char_vector) {
    i <- 1
    result <- c()
    while (i <= length(char_vector)) {
      if (char_vector[i] == "^") {
        i <- i + 2
      } else {
        result <- c(result, char_vector[i])
        i <- i + 1
      }
    }
    result
  }

  remove_dollar_sign <- function(char_vector) {
    char_vector[char_vector != "$"]
  }

  create_alignment_matrix <- function(positions, 
                                       base_lists, 
                                       readnames_lists, 
                                       repeated_reads, 
                                       remove_repeated_reads = FALSE) {
    all_readnames <- unique(unlist(readnames_lists))
    if (remove_repeated_reads) {
      all_readnames <- setdiff(all_readnames, repeated_reads)
    }
    alignment_matrix <- matrix(NA, nrow = length(all_readnames), ncol = length(positions),
                               dimnames = list(all_readnames, positions))
    for (i in seq_along(positions)) {
      position <- positions[i]
      base_info <- base_lists[[i]]
      readnames <- readnames_lists[[i]]
      for (j in seq_along(readnames)) {
        if (!remove_repeated_reads || !(readnames[j] %in% repeated_reads)) {
          alignment_matrix[readnames[j], as.character(position)] <- base_info[j]
        }
      }
    }
    alignment_matrix[is.na(alignment_matrix)] <- ""
    return(alignment_matrix)
  }

  adjust_matrix_with_reference <- function(matrix, refpos_df) {
    for (position in colnames(matrix)) {
      numeric_position <- as.numeric(position)
      ref_nucleotide <- refpos_df$REF[refpos_df$POS == numeric_position]
      if (length(ref_nucleotide) > 0) {
        matrix[, position][matrix[, position] %in% c(",", ".")] <- ref_nucleotide
      }
    }
    return(matrix)
  }

  capitalize_bases <- function(matrix) {
    for (i in seq_len(nrow(matrix))) {
      for (j in seq_len(ncol(matrix))) {
        value <- matrix[i, j]
        if (value %in% c("a", "t", "g", "c")) {
          matrix[i, j] <- toupper(value)
        }
      }
    }
    return(matrix)
  }

  # Hauptprozess
  pileup_data <- fread(vcf_file_path, sep = "\t", header = FALSE, quote = "", fill = TRUE)
  colnames(pileup_data) <- c("Chromosome", "Position", "ReferenceBase", "Coverage", "ReadBases", "Qualities", "noIdea", "ReadNames")
  
  positions <- pileup_data$Position
  base_lists <- list()
  readnames_lists <- list()
  duplicated_read_names <- c()
  
  for (i in seq_len(nrow(pileup_data))) {
    base_string <- pileup_data$ReadBases[i]
    char_vector <- split_string_to_chars(base_string)
    result <- process_plus_minus_in_vector(char_vector)
    result <- replace_asterisk_and_hash(result)
    result <- remove_caret_and_next(result)
    result <- remove_dollar_sign(result)
    base_lists[[i]] <- result
    
    readnames <- unlist(strsplit(pileup_data$ReadNames[i], ","))
    readnames_lists[[i]] <- readnames
    readname_counts <- table(readnames)
    repeated_reads <- names(readname_counts[readname_counts > 1])
    if (length(repeated_reads) > 0) {
      duplicated_read_names <- unique(c(duplicated_read_names, repeated_reads))
    }
  }
  
  result_matrix <- create_alignment_matrix(positions, base_lists, 
                                           readnames_lists, duplicated_read_names, 
                                           remove_repeated_reads = TRUE)
  
  RefPos <- data.frame(REF = pileup_data$ReferenceBase, POS = pileup_data$Position)
  adjusted_matrix <- adjust_matrix_with_reference(result_matrix, RefPos)
  adjusted_matrix <- capitalize_bases(adjusted_matrix)
  
  return(adjusted_matrix)
}
```

## --------

## Generating SNP matrix

```{r}
# Funktion definieren, um SNP-Matrix zu erstellen, zu speichern und zurückzugeben
process_and_save_snp_matrices <- function(file_paths) {
  snp_matrices <- list()
  
  for (file_path in file_paths) {
    file_name <- sub("\\.pileup$", "", basename(file_path))
    output_file <- paste0("snp_matrix/snp_matrix_", file_name, ".txt")
    
    # SNP-Matrix erstellen
    snp_matrix <- process_vcf_to_snp_matrix(file_path)
    
    # Matrix speichern
    save_matrix <- snp_matrix
    save_matrix[save_matrix == ""] <- "X"
    write.table(save_matrix, file = output_file, 
                sep = "\t", col.names = TRUE, row.names = TRUE, quote = FALSE)
    
    # Matrix in die Liste speichern
    snp_matrices[[file_name]] <- snp_matrix
  }
  
  return(snp_matrices)
}

# Aufruf der Funktion für mehrere Dateien
file_paths <- c("pileup/SRR26992682.pileup", "pileup/SRR26992684.pileup")
snp_matrices <- process_and_save_snp_matrices(file_paths)

snp_matrices[[1]][1:5, 1:20]
```


## --------

## Position Weight Matrix 

```{r}
getPWM = function( seqs, sym=c("A","C","G","T","-") ) {
	t(apply(seqs, 2, function(nuc) {
		idx=which(nuc!="")
		t=table(nuc[idx])
		d=t[sym]
		d[which(is.na(d))] = 0
		names(d)=sym;
		return(d/sum(t))
	} ))
}

pwm_list <- list()

pwm_list <- list()
for (i in seq_along(snp_matrices)) {
  seqs <- snp_matrices[[i]]
  
  pwm <- getPWM(seqs)
  pwm_list[[names(snp_matrices)[i]]] <- pwm
}

head(pwm_list[["SRR26992684"]])
#head(pwm_list[["SRR26992682"]])
```
### PWM frequency sorted

```{r}
# Schritt 2: PWMs sortieren und in separater Liste speichern
pwm_list_sorted <- list()
for (i in seq_along(pwm_list)) {
  pwm <- pwm_list[[i]]
  pwm_sorted <- t(apply(pwm, 1, sort, decreasing = TRUE))
  pwm_list_sorted[[names(pwm_list)[i]]] <- pwm_sorted
}

#head(pwm_list_sorted[["SRR26992684"]])
head(pwm_list_sorted[["SRR26992684"]])
```

## Empirical cumulative distribution function (ECDF)

```{r}
ecdf_plot <- function(pwm){
  
  p <- ggplot() + 
    geom_vline(xintercept = 0, linetype='dashed', color='grey') +
    geom_vline(xintercept = 1, linetype='dashed', color='grey') +
    geom_hline(yintercept = 0, linetype='dashed', color='grey') +
    geom_hline(yintercept = 1, linetype='dashed', color='grey') +
    stat_ecdf(aes(pwm[,4]), col="blue") +
    stat_ecdf(aes(pwm[,3]), col="green") +
    stat_ecdf(aes(pwm[,2]), col="red") +
    stat_ecdf(aes(pwm[,1]), col="black") +
    theme(
      axis.text.x = element_text(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(colour = "black"),
      panel.background = element_rect(fill = "white"),
      panel.border = element_rect(colour = "black", fill= NA, linewidth = 1),
      legend.position = "none"
    ) +
    theme(
      strip.background = element_blank(),
      strip.text.x = element_blank()
    ) +
    xlab("Relative nucleotide frequency") +
    ylab("% of SNV positions")
  
  return(p)
}


p1 <- ecdf_plot(pwm_list_sorted[["SRR26992684"]])
p2 <- ecdf_plot(pwm_list_sorted[["SRR26992682"]])

## combined plot with patchwork
combined_plot <- (p1 | p2) +
  plot_annotation(title = '') +
  plot_annotation(tag_levels = 'A')

print(combined_plot)

f <- 1

w <- 7
h <- 3.5

ggsave("figures/ecdf.png", plot = combined_plot, width = w * f, height = h * f)
```

## --------

## Filter reads by covered SNP positions

A read should have at least 10 SNP positions.

```{r}
# Funktion zum Zählen leerer Felder
fun_empty <- function(x) {
  sum(x == "" | is.na(x))  # Berücksichtigt leere Werte und NA
}

# Funktion zum Zählen von Nicht-Leerwerten
fun_not_empty <- function(x) {
  sum(x != "" & !is.na(x))  # Zählt alle Werte, die nicht leer oder NA sind
}

# Anwenden der Funktionen auf die Zeilen der Matrix
empty <- apply(adjusted_matrix, 1, fun_empty)
notEmpty <- apply(adjusted_matrix, 1, fun_not_empty)
#Reads with more than 10 SNPs
positionsCovered <- notEmpty-empty
seqsFiltered1 <- adjusted_matrix[which(positionsCovered >= 10), ]
```



## Filter by second allel frequency

```{r}
getPWM = function( seqs, sym=c("A","C","G","T","-") ) {
	t(apply(seqs,2,function(nuc) {
		idx=which(nuc!="")
		t=table(nuc[idx])
		d=t[sym]
		d[which(is.na(d))]= 0
		names(d)=sym;
		return(d/sum(t))
	} ))
}

pwm <- getPWM(seqsFiltered1)

pwm <- t(apply(pwm, 1, sort, decreasing = TRUE))

seqsFiltered2 <- seqsFiltered1[, which(pwm[, 2] >= 0.1)]

head(pwm)
```







